[DOC-9249] Add HASHBYTES Function (#327)
couchbaselabs/docs-devex: modules/n1ql/pages/n1ql-language-reference/metafun.adoc

----
====

[[hashbytes,HASHBYTES()]]

== HASHBYTES(`input`, [ `options` ])

=== Description

This function returns a binary hash value for a given input using a specified hashing algorithm. 

> By using this function, you can protect your data by masking its original form while still allowing verification or comparison. 

Reviewer: Just tweaking the wording to emphasize that it's not just for security.

```suggestion
By using this function, you can verify or compare data quickly, or protect your data by masking its original form while still allowing verification or comparison. 
```

=== Arguments

> input:: The data to be hashed, represented as a binary object.
> +
> If the data is not in the binary form, you must convert it to a JSON marshalled value (JSON string) before it can be hashed. 
> You can use the ENCODE_JSON() function to do this conversion. 
> For more information, see xref:n1ql-language-reference/jsonfun.adoc[JSON Functions].     

Reviewer: Looking closely at the original ticket, I'm not sure if this is quite right.

```suggestion
input:: A binary object or any {sqlpp} data type.
The JSON marshalled value of the data is used as the input.
```

That's how I read it anyway. If you're in doubt you can test the input by trying out the function with a number, an object, an array, NULL, and MISSING, and see whether you get valid output each time.

> options:: An object that specifies the hashing algorithm (string) to be used.
> +
> Supported algorithms are: `crc32`, `md4`, `md5`, `sha224`, `sha256`, `sha384`, `sha512`, `sha512/224`, and `sha512/256`. 
> +
> Some of the algorithms require additional configuration fields. 
> For example, for `crc32`, you need to add a `polynomial` field which can have one of the following values: 
> + 
> --
> * `ieee` (the default value)
> * `castagnoli`
> * `koopman`
> * A valid 32-bit integer, provided either as a JSON number (decimal) or a string that can be parsed as a numeric value (supports hexadecimal with a "0x" prefix)
> --

Reviewer: We need to be more explicit in the way the options object is documented. For an example of how we do this, see [_TIMESERIES Function](https://docs.couchbase.com/server/current/n1ql/n1ql-language-reference/timeseries.html).  Here's how I would do it:

```suggestion
options::  [Optional]
An object that specifies the hashing algorithm and other options for the function.
If omitted, the default hashing algorithm is `sha256`.

=== Options

[options="header", cols="1a,3a,1a"]
|===
| Name | Description | Schema

| **algorithm** +
__required__
| Specifies the hashing algorithm to be used.

Supported algorithms are: `crc32`, `md4`, `md5`, `sha224`, `sha256`, `sha384`, `sha512`, `sha512/224`, and `sha512/256`. 
| String

| **polynomial** +
__optional__
| The polynomial to use.
This only applies if the algorithm is `crc32`.
This property may have one of the following values:

* `ieee` (the default value)
* `castagnoli`
* `koopman`
* A valid 32-bit integer, provided either as a JSON number (decimal) or a string that can be parsed as a numeric value (supports hexadecimal with a "0x" prefix)

*Default:* `ieee`
| String or integer
|===
```
